src/nvim/autocmd.c

```diff
@@ -67,28 +83,54 @@
 // Code for automatic commands.
 static AutoPatCmd *active_apc_list = NULL;  // stack of active autocommands
 
-/// List of autocmd group names
-static garray_T augroups = { 0, 0, sizeof(char_u *), 10, NULL };
-#define AUGROUP_NAME(i) (((char **)augroups.ga_data)[i])
-#define BUFLOCAL_PAT_LEN 25
+// ID for associating autocmds created via nvim_create_autocmd
+// Used to delete autocmds from nvim_del_autocmd
+static int next_augroup_id = 1;
 
 // use get_deleted_augroup() to get this
 static const char *deleted_augroup = NULL;
 
-// The ID of the current group.  Group 0 is the default one.
+// The ID of the current group.
 static int current_augroup = AUGROUP_DEFAULT;
 
-static int au_need_clean = false;  // need to delete marked patterns
+// Whether we need to delete marked patterns.
+// While deleting autocmds, they aren't actually remover, just marked.
+static int au_need_clean = false;
 
-static event_T last_event;
-static int last_group;
 static int autocmd_blocked = 0;  // block all autocmds
 
 static bool autocmd_nested = false;
 static bool autocmd_include_groups = false;
 
 static char_u *old_termresponse = NULL;
 
+/// Iterates over all the AutoPats for a particular event
+#define FOR_ALL_AUPATS_IN_EVENT(event, ap) \
+  for (AutoPat *ap = first_autopat[event]; ap != NULL; ap = ap->next)  // NOLINT
+
+/// Handles grabbing arguments from `:autocmd` such as ++once and ++nested
+#define ARG_GET_FLAG(errored, cmd, flag, pattern, len) \
```

Comment:
not sure if factoring out the three usecases is worth it, but if it is, it could be a function taking `flag` and `cmd` as pointers.